File: ASD/intro_trees/iter_postorder.c
#include <stdlib.h>
#include <stdio.h>

#include "iter_postorder.h"
#include "stack.h"
#include "tree.h"

void _iterative_postorder_visit(node_tree* node, int depth) {
    node_t* n;
    for (n = (node_t*) node; n != NULL; n = (node_t*) n->sibling) {
        _iterative_postorder_visit(n->child, depth + 1);
        print_depth(depth);
        printf("%d\n", n->info);
    }
}

void iterative_postorder_visit(node_tree * node) {
    _iterative_postorder_visit(node, 0);
}

File: ASD/Exams/20150914/problem_2/java-sol/IterPostOrder.java
public class IterPostOrder {

    public static void iterativePostOrderVisit(NodeTree node){

        Stack s = new Stack();
        boolean from_stack = false;

        while(s.size() > 0 || node != null) {

            if (node != null) {

                if (from_stack == false && node.getFirstChild() != null) {

                    s.push(node);
                    node = node.getFirstChild();

                } else {

                    System.out.println("Visiting node: " + node.getNodeInfo());
                    node = node.getNextSibling();

                }

            }

            if (node == null) {

                if (s.size() > 0) {
                    node = s.pop();
                    from_stack = true;
                }

            } else {
                from_stack = false;
            }
        }
    }

}
